core/src/state/binding.rs

```diff
@@ -1,141 +1,264 @@
+    pub store_id: TypeId,
+    pub model_id: TypeId,
+}
+
+pub trait Bindable: 'static + Clone {
+    type Output;
+    fn view<D: DataContext, F: FnOnce(Option<&Self::Output>) -> T, T>(
+        &self,
+        cx: &D,
+        viewer: F,
+    ) -> T;
+    fn get<D: DataContext>(&self, cx: &D) -> Option<Self::Output>
     where
-        F: 'static + Fn(&mut Context, L),
+        Self::Output: Clone,
     {
-        let binding = Self { lens: lens.clone(), content: Some(Box::new(builder)) };
-
-        let id = cx.entity_manager.create();
-        let current = cx.current();
-        cx.tree().add(id, current).expect("Failed to add to tree");
-        cx.cache().add(id).expect("Failed to add to cache");
-        cx.style().add(id);
-
-        let ancestors = cx.current().parent_iter(cx.tree()).collect::<HashSet<_>>();
-        let new_ancestors = id.parent_iter(cx.tree()).collect::<Vec<_>>();
-
-        for entity in new_ancestors {
-            if let Some(model_data_store) = cx.data.get_mut(entity) {
-                if let Some(model_data) = model_data_store.data.get(&TypeId::of::<L::Source>()) {
-                    if let Some(lens_wrap) =
-                        lens.cache_key().and_then(|key| model_data_store.lenses_dedup.get_mut(&key))
-                    {
-                        let observers = lens_wrap.observers();
-
-                        if ancestors.intersection(observers).next().is_none() {
-                            lens_wrap.add_observer(id);
-                        }
-                    } else {
-                        let mut observers = HashSet::new();
-                        observers.insert(id);
+        self.view(cx, |v| v.cloned())
+    }
 
-                        let model = model_data.downcast_ref::<L::Source>().unwrap();
+    fn requests(&self) -> Vec<StoreId>;
+    fn make_store(&self, source: ModelOrView) -> Option<Box<dyn StoreHandler>>;
+    fn add_to_store(&self, store: &mut dyn StoreHandler, entity: Entity);
+}
 
-                        let old = lens.view(model, |t| t.cloned());
+impl<L> Bindable for L
+where
+    L: Lens,
+    <L as Lens>::Source: 'static,
+    <L as Lens>::Target: Data,
+{
+    type Output = L::Target;
 
-                        let state = Box::new(State { entity: id, lens, old, observers });
+    fn view<D: DataContext, F: FnOnce(Option<&Self::Output>) -> T, T>(
+        &self,
+        cx: &D,
+        viewer: F,
+    ) -> T {
+        self.view(
+            cx.data().expect(
+                "Lens failed to retrieve data from context. Did you forget to build a model?",
+            ),
+            viewer,
+        )
+    }
 
-                        if let Some(key) = state.lens.cache_key() {
-                            model_data_store.lenses_dedup.insert(key, state);
-                        } else {
-                            model_data_store.lenses_dup.push(state);
-                        }
-                    }
+    fn requests(&self) -> Vec<StoreId> {
```

Comment:
I'm a bit confused how this works because the `make_store` method seems to only be able to create a single store but this method suggests that a bindable can link to many stores.

Reply:
make_store will be called once for each request!

Reply:
Ah okay, so the requests vector defines what stores the binable links to and then the `make_store` functions creates the store for each item?

Reply:
Correct